server_cached_router.go

```diff
@@ -0,0 +1,247 @@
+// It's a bit complex because it ensures we continue iterating without blocking on the FindPeers call.
+func NewCacheFallbackIter(sourceIter iter.ResultIter[types.Record], router cachedRouter, ctx context.Context) *cacheFallbackIter {
+	ctx, cancel := context.WithCancel(ctx)
+	return &cacheFallbackIter{
+		sourceIter:      sourceIter,
+		router:          router,
+		ctx:             ctx,
+		cancel:          cancel,
+		findPeersResult: make(chan types.PeerRecord),
+		ongoingLookups:  atomic.Int32{},
+	}
+}
+
+func (it *cacheFallbackIter) Next() bool {
+	// Try to get the next value from the source iterator first
+	if it.sourceIter.Next() {
+		val := it.sourceIter.Val()
+		handleRecord := func(id *peer.ID, record *types.PeerRecord) bool {
+			record.Addrs = it.router.withAddrsFromCache(addrQueryOriginProviders, *id, record.Addrs)
+			if record.Addrs != nil {
+				it.current = iter.Result[types.Record]{Val: record}
+				return true
+			}
+			logger.Infow("no cached addresses found in cacheFallbackIter, dispatching find peers", "peer", id)
+
+			if it.router.cachedAddrBook.ShouldProbePeer(*id) {
+				it.ongoingLookups.Add(1) // important to increment before dispatchFindPeer
+				// If a record has no addrs, we dispatch a lookup to find addresses
+				go it.dispatchFindPeer(*record)
+
+				return it.Next() // Recursively call Next() to either read from sourceIter or wait for lookup result
+			}
+
+			// If we're not going to probe, return the record with no addrs
+			// TODO: should we even return these if the peer is likely unreachable?
+			it.current = iter.Result[types.Record]{Val: record}
+			return true
+		}
+
+		switch val.Val.GetSchema() {
+		case types.SchemaPeer:
+			if record, ok := val.Val.(*types.PeerRecord); ok {
+				return handleRecord(record.ID, record)
+			}
+		}
+		it.current = val // pass through unknown schemas
+		return true
+	}
+
+	// If there are still ongoing lookups, wait for them
+	if it.ongoingLookups.Load() > 0 {
+		logger.Infow("waiting for ongoing find peers result")
+		select {
+		case result, ok := <-it.findPeersResult:
+			if ok {
+				it.current = iter.Result[types.Record]{Val: &result}
+				return true
+			}
+		case <-it.ctx.Done():
+			return false
+		}
+	}
+
+	return false
+}
+
+func (it *cacheFallbackIter) Val() iter.Result[types.Record] {
+	if it.current.Val != nil || it.current.Err != nil {
+		return it.current
+	}
+	return iter.Result[types.Record]{Err: errNoValueAvailable}
+}
+
+func (it *cacheFallbackIter) dispatchFindPeer(record types.PeerRecord) {
+	defer it.ongoingLookups.Add(-1)
+	// FindPeers is weird in that it accepts a limit. But we only want one result, ideally from the libp2p router.
+	peersIt, err := it.router.FindPeers(it.ctx, *record.ID, 1)
+
+	if err != nil {
+		it.findPeersResult <- record // pass back the record with no addrs
```

Comment:
@aschmahmann noted that if we made probe and fail to find addrss for that PeerID, there is no point returning that PeerID without addrs, because client will try to do the same lookup (potentially hitting /routing/v1/peers too), and fail again. 

So let's skip PeerIDs that have no addrs and fail FindPeers here.

Reply:
Fixed in https://github.com/ipfs/someguy/pull/90/commits/c5abeec8c18055d9c3c871db2f387ac819f38e27

Reply:
@lidel Should we also omit cached addresses from `FindPeer` requests? Currently, if the libp2p `FindPeer` fails, we try returning from the cache. But maybe that's a bad idea? 

https://github.com/ipfs/someguy/blob/c5abeec8c18055d9c3c871db2f387ac819f38e27/server_cached_router.go#L70-L99

Reply:
As discussed in person with @lidel. We will avoid looking up the cache in FindPeer queries, since the FindPeer should always return whether a peer is currently dialable. 

We will only record connection failures for probing purposes.